app/assets/javascripts/timeout.js

```diff
@@ -0,0 +1,149 @@
+    if (displayInactivityPopup()) {
+      inactivity_dialog = setTimeout(function () {
+        if (checkLocalStorage()) {
+          lastActiveTimeout();
+        } else {
+          const inactivity_popup = document.getElementById("activity");
+          inactivity_popup.showModal();
+          startInactivityPopupTimeout();
+        }
+      }, 1000 * 60 * inactivity_mins);
+    }
+  }
+
+  function resetTimeouts() {
+    // If stay signed in button is clicked then activity timeout is restarted
+    const staySignedInButton = document.getElementById(
+      "hmrc-timeout-keep-signin-btn"
+    );
+    const inactivity_popup = document.getElementById("activity");
+    if (staySignedInButton) {
+      staySignedInButton.addEventListener("click", function () {
+        clearTimeout(inactivity_dialog);
+        clearTimeout(inactivity_logout);
+        inactivity_popup.close();
+        startInactivityTimeout();
+      });
+    }
+  }
+
+  function startInactivityPopupTimeout() {
+    const inactivity_popup = document.getElementById("activity");
+    // Logs user out if no response when the inactivity popup shows
+    inactivity_logout = setTimeout(function () {
+      inactivity_popup.close();
+      signOutRedirect();
+    }, 1000 * 60 * initial_warning_mins);
+  }
+
+  function sessionExpiryPopup() {
+    // displays session expiry popup after timeout
+    const session_expiry_popup = document.querySelector("#expiry");
+    if (isLoggedIn()) {
+      inactivity_logout = setTimeout(function () {
+        if (session_expiry_popup)
+        {
+          session_expiry_popup.showModal();
+        }
+      }, getTimeLeftUntilExpiry());
+    }
+  }
+
+  function updateLocalStorage() {
+    // With each request in any tab, the lastActive attribute is updated
+    localStorage.setItem("lastActivity", new Date());
+  }
+
+  function checkLocalStorage() {
+    // Checking local storage for any activity in other tabs
+    let lastActive = new Date(localStorage.getItem("lastActivity"));
+    let time = new Date();
+    return (time - lastActive) < (inactivity_mins * 60 * 1000);
+  }
+
+  function lastActiveTimeout() {
+    // If activity in another tab, inactivity timeout period adjusted
+    const inactivity_popup = document.getElementById("activity");
+    let time = new Date();
+    let lastActive = new Date(localStorage.getItem("lastActivity"));
+    setTimeout(() => {
+      inactivity_popup.showModal();
+      startInactivityPopupTimeout();
+    }, inactivity_mins * 60 * 1000 - (time - lastActive));
+  }
+
+  function signOutRedirect() {
+    // send logout request and redirect to previous page upon relogging in
+    let current_page = window.location.pathname;
+    $.ajax("/sign-out", {
+      method: "GET",
+      success: function() {
+        window.location.href = "/sign-in?next="+current_page+"&status=inactive";
+      }
+    })
+      .fail((response) => {
+        console.log(response);
+      });
```

Comment:
Any particular reason you're using promises for fail and callbacks for success?

What happens if current_page contains characters which need URI escaping and/or it is a page which has been POSTed to or there are query string parameters?